Cooperate with custom error handlers

DiscordLiz opened the pull request:
Adds detection of custom error handlers to the global error handler.

Adds documentation of the updated behavior.

closes #2276

### Type

- [x] Bugfix
- [ ] Enhancement
- [ ] New feature

### Description of the changes

Commit:
Cooperate with custom error handlers

Adds detection of custom error handlers to the global error handler.

Adds documentation of the updated behavior.

closes #2276

calebj reviewed (changes requested):
It looks like Red already ignores (or at least, doesn't post) errors if a cog-wide exception handler is present (see line 191). If that becomes unreachable code, better to just skip that check altogether.

Commit:
Address feedback on github

Commit:
Handle feedback brought up in discord.

DiscordLiz commented:
I handled the feedback brought up on Discord and on GitHub separately.

@calebj brought up in discord that cog authors may not want to handle every error red does themselves.

I added a keyword only argument in the latest commit to allow developers to pass their errors back to Red's handler and have it skip the checks for if there were a cog handler already.

Other solutions to this which were discussed in discord.

1. Split error handling up into another function and the event. (@mikeshardmind brought this up)

This would incure an additional function call per error which wasn't handled, but would leave the event signature untouched.

2. Move error handling into the cog base class. (@mikeshardmind brought this up)

This would require significant changes to error handling to select the first handler in a cog class's mro, but would allow passing of unhandled exceptions using super.

3. Create a new convention of a red specific error handler for each type of error. (@calebj brought this up)

This would put an additional learning curve to new developers transitioning from using discord.py without Red, and would still require significant changes to upstream code to prevent double error handling.

This would also not allow people to use a cog error handler to log exceptions broadly without duplicating code.

calebj reviewed (changes requested):
Is it `manually_recalled` or `manually_reinvoked`?

Commit:
Use a more meaningful name for kwarg

Tobotimus commented on redbot/core/commands/commands.py:
```diff
@@ -398,6 +398,44 @@ def clear_rule_for(
+-------
+discord.ClientException
+    The coroutine is not actually a coroutine.
+"""
```
Could you please create the `error(self, coro)` method on our command class to just call `super().error(coro)`, and then move this docstring there? That way it'll show up in our docs, and is more obvious in the code.

DiscordLiz replied:
I've created the method and moved the docstring.

Tobotimus reviewed (changes requested):
Just one change I'm requesting then it's good :)

Commit:
Allow discovery of docstring by sphinx.

Commit:
remove invalid reference

Tobotimus committed:
Docstring markdown
